### ISSUE TITLE
In AUTO_ACK mode message is immediately nack when exception occurs.

### ISSUE THREAD

**Issue by a-wrobel** [spring-cloud/spring-cloud-gcp#2250]
**According to AckMode.AUTO_ACK docs:**

The framework acks the {@link com.google.pubsub.v1.PubsubMessage} after it is
sent to the channel and processed successfully.
<p>The framework does not immediately nack the message when the exception occurs,
and allows the eventual redelivery to take effect.

**Actual behaviour:**

The message is immediately redelivered when the exception occurs.
For any exception PubSubInboundChannelAdapter throws PubSubException, 
due to this message is nacked MessageDispatcher.AckHandler in onFailure.

**Comment by elefeint** [comment]
You are correct; the documentation is wrong.

In order to get the neither-acking-nor-nacking behavior, set up a Spring Integration error channel, like so:
````
@Bean
public MessageChannel pubsubInputChannel() {
    return new PublishSubscribeChannel();
}

@Bean
public PubSubInboundChannelAdapter messageChannelAdapter(
    @Qualifier("pubsubInputChannel") MessageChannel inputChannel,
    SubscriberFactory subscriberFactory) {
    PubSubInboundChannelAdapter adapter =
        new PubSubInboundChannelAdapter(subscriberFactory, "subscriptionName");
    adapter.setOutputChannel(inputChannel);
    adapter.setAckMode(AckMode.AUTO_ACK);

    // Error redirection, to avoid the client library acking by default:
    adapter.setErrorChannelName("pubsubErrors");

    return adapter;
}

@ServiceActivator(inputChannel =  "pubsubErrors")
public void pubsubErrorHandler(Message<MessagingException> message) {
	LOGGER.warn("This message is neither acked nor nacked; ack deadline will determine redelivery");
}
````

For our reference, what version of the Pub/Sub starter are you using?

**Comment by meltsufin** [comment]
Documentation fix is pending here: https://github.com/spring-cloud/spring-cloud-gcp/pull/2078

**Comment by a-wrobel** [comment]
Sadly after apply a error channel message has not been redelivered. I have used 1.2.2.RELEASE version.

**Comment by meltsufin** [comment]
@a-wrobel How long did you wait? Redelivery can happen up to 10 minutes later or so.

**Comment by elefeint** [comment]
I'll investigate in the afternoon. I remember looking into it with Dan, and with error channel the default acking behavior got triggered (because there was no exception).
What I don't remember is whether we realized that before or after I wrote that documentation PR.

**Comment by a-wrobel** [comment]
A message was not redelivered for 15 min. For testSubscrition I have set 60sec ack deadline.
A sample code can be found under https://github.com/a-wrobel/pub-sub.

**Comment by elefeint** [comment]
@dhoard when you originally implemented AUTO_ACK in Spring Cloud GCP with Spring Integration in #856, did it behave as you expected? Or did you create googleapis/java-pubsub#20 because it was still not possible to set up "message neither acked nor nacked, and instead allowed to expire and get redelivered" logic?

Looking at the code, there are two possible uses of AUTO_ACK when the downstream processing errors out:
1) without a custom error handler, error is propagated to `PubSubInboundChannelAdapter`. The adapter neither nacks nor acks, however the exception is [rethrown](https://github.com/spring-cloud/spring-cloud-gcp/blob/256f9a99f5abfea662197c153bc45a8b600f018f/spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/integration/inbound/PubSubInboundChannelAdapter.java#L162) up to the client library, causing the client library to nack the message.

2) with a custom error handler that does not propagate the exception, no error is propagated, and the message is [acked](https://github.com/spring-cloud/spring-cloud-gcp/blob/256f9a99f5abfea662197c153bc45a8b600f018f/spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/integration/inbound/PubSubInboundChannelAdapter.java#L155).

The simplest solution, as suggested in #2074, is to avoid rethrowing the exception in AUTO_ACK and MANUAL modes, logging it as a warning instead.

I wanted to get your thoughts on the issue first, though.

**Comment by dhoard** [comment]
@elefeint ... haven't used GCP Pub/Sub in over a year/no longer at the company using the calling code.

https://github.com/spring-cloud/spring-cloud-gcp/pull/856 was oriented around the use of  Spring / `PubSubInboundChannelAdapter`

https://github.com/googleapis/java-pubsub/issues/20 was oriented around the use of the core GCP Pub/Sub client.

I don't remember any issues... but we used a combination of spring-cloud-gcp, the core GCP Pub/Sub client library, and a custom Pub/Sub client library... depending on when the code was written and the issues we experienced.

I feel this is a bug. I think the code concept in https://github.com/spring-cloud/spring-cloud-gcp/issues/2074 is correct to resolve the issue. This would be consistent with the documentation...

Automatic acking OK (AckMode.AUTO_ACK)

"A message is acked with GCP Pub/Sub if the adapter sent it to the channel and no exceptions were thrown. If a RuntimeException is thrown while the message is processed, then the message is neither acked / nor backed."
